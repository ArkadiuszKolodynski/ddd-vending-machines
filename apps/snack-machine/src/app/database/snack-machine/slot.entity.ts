import { Cascade, Embeddable, Embedded, Entity, ManyToOne, OneToOne, PrimaryKey, Property } from '@mikro-orm/core';
import SnackEntity from '../snack/snack.entity';
import SnackMachineEntity from './snack-machine.entity';

@Embeddable()
export class SnackPileEmbeddable {
  @Property({ type: 'int' })
  quantity!: number;

  @Property({ type: 'varchar', length: 255 })
  price!: string;

  @OneToOne({ cascade: [Cascade.ALL], unique: false })
  snack!: SnackEntity;
}

@Entity({ tableName: 'slot' })
export default class SlotEntity {
  @PrimaryKey({ type: 'uuid' })
  id!: string;

  @Property({ type: 'int' })
  position!: number;

  @ManyToOne(() => SnackMachineEntity)
  snackMachine!: SnackMachineEntity;

  @Embedded(() => SnackPileEmbeddable)
  snackPile!: SnackPileEmbeddable;
}
