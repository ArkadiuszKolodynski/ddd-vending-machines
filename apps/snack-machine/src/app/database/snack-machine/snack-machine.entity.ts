import {
  BaseEntity,
  Cascade,
  Collection,
  Embeddable,
  Embedded,
  Entity,
  OneToMany,
  PrimaryKey,
  Property,
} from '@mikro-orm/core';
import SlotEntity from './slot.entity';

@Embeddable()
export class MoneyEmbeddable {
  @Property({ type: 'int' })
  oneCentCount!: number;

  @Property({ type: 'int' })
  tenCentCount!: number;

  @Property({ type: 'int' })
  quarterCount!: number;

  @Property({ type: 'int' })
  oneDollarCount!: number;

  @Property({ type: 'int' })
  fiveDollarCount!: number;

  @Property({ type: 'int' })
  twentyDollarCount!: number;
}

@Entity({ tableName: 'snack_machine' })
export default class SnackMachineEntity extends BaseEntity {
  @PrimaryKey({ type: 'uuid' })
  id!: string;

  @Embedded(() => MoneyEmbeddable)
  money!: MoneyEmbeddable;

  @Property({ type: 'varchar', length: 255 })
  moneyInTransaction!: string;

  @OneToMany(() => SlotEntity, (slot) => slot.snackMachine, { cascade: [Cascade.ALL], orphanRemoval: true })
  slots = new Collection<SlotEntity>(this);
}
