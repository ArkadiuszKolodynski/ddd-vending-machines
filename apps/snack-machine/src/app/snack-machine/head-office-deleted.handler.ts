import { CreateRequestContext, MikroORM } from '@mikro-orm/core';
import { Logger } from '@nestjs/common';
import { EventPublisher, EventsHandler, IEventHandler } from '@nestjs/cqrs';
import { HeadOfficeDeletedEvent } from '@vending-machines/events';
import { SnackMachineRepository } from './snack-machine.repository.interface';

@EventsHandler(HeadOfficeDeletedEvent)
export class HeadOfficeDeletedHandler implements IEventHandler<HeadOfficeDeletedEvent> {
  constructor(
    private readonly orm: MikroORM, // MikroORM needed for @CreateRequestContext()
    private readonly snackMachineRepository: SnackMachineRepository,
    private readonly eventPublisher: EventPublisher,
  ) {}

  @CreateRequestContext()
  async handle(event: HeadOfficeDeletedEvent) {
    const existingSnackMachine = await this.snackMachineRepository.findOne(event.aggregateId);
    if (!existingSnackMachine) {
      Logger.error(`Head office with id ${event.aggregateId} not found`, this.constructor.name);
      return;
    }

    const snackMachine = this.eventPublisher.mergeObjectContext(existingSnackMachine);
    snackMachine.markAsDeleted();
    await this.snackMachineRepository.delete(snackMachine.id);
    snackMachine.commit();
  }
}
