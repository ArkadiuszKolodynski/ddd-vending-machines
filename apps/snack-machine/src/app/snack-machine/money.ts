import { ValueObject } from '@vending-machines/shared';
import currency from 'currency.js';

export class Money extends ValueObject {
  static readonly None = new Money(0, 0, 0, 0, 0, 0);
  static readonly Cent = new Money(1, 0, 0, 0, 0, 0);
  static readonly TenCent = new Money(0, 1, 0, 0, 0, 0);
  static readonly Quarter = new Money(0, 0, 1, 0, 0, 0);
  static readonly Dollar = new Money(0, 0, 0, 1, 0, 0);
  static readonly FiveDollar = new Money(0, 0, 0, 0, 1, 0);
  static readonly TwentyDollar = new Money(0, 0, 0, 0, 0, 1);

  readonly oneCentCount: number;
  readonly tenCentCount: number;
  readonly quarterCount: number;
  readonly oneDollarCount: number;
  readonly fiveDollarCount: number;
  readonly twentyDollarCount: number;

  get amount(): currency {
    return new currency(this.oneCentCount)
      .multiply(0.01)
      .add(new currency(this.tenCentCount).multiply(0.1))
      .add(new currency(this.quarterCount).multiply(0.25))
      .add(new currency(this.oneDollarCount).multiply(1))
      .add(new currency(this.fiveDollarCount).multiply(5))
      .add(new currency(this.twentyDollarCount).multiply(20));
  }

  constructor(
    oneCentCount: number,
    tenCentCount: number,
    quarterCount: number,
    oneDollarCount: number,
    fiveDollarCount: number,
    twentyDollarCount: number,
  ) {
    super();

    if (oneCentCount < 0) {
      throw new Error('Invalid operation');
    }
    if (tenCentCount < 0) {
      throw new Error('Invalid operation');
    }
    if (quarterCount < 0) {
      throw new Error('Invalid operation');
    }
    if (oneDollarCount < 0) {
      throw new Error('Invalid operation');
    }
    if (fiveDollarCount < 0) {
      throw new Error('Invalid operation');
    }
    if (twentyDollarCount < 0) {
      throw new Error('Invalid operation');
    }

    this.oneCentCount = oneCentCount;
    this.tenCentCount = tenCentCount;
    this.quarterCount = quarterCount;
    this.oneDollarCount = oneDollarCount;
    this.fiveDollarCount = fiveDollarCount;
    this.twentyDollarCount = twentyDollarCount;
  }

  static add(money1: Money, money2: Money): Money {
    return new Money(
      money1.oneCentCount + money2.oneCentCount,
      money1.tenCentCount + money2.tenCentCount,
      money1.quarterCount + money2.quarterCount,
      money1.oneDollarCount + money2.oneDollarCount,
      money1.fiveDollarCount + money2.fiveDollarCount,
      money1.twentyDollarCount + money2.twentyDollarCount,
    );
  }

  static subtract(money1: Money, money2: Money): Money {
    return new Money(
      money1.oneCentCount - money2.oneCentCount,
      money1.tenCentCount - money2.tenCentCount,
      money1.quarterCount - money2.quarterCount,
      money1.oneDollarCount - money2.oneDollarCount,
      money1.fiveDollarCount - money2.fiveDollarCount,
      money1.twentyDollarCount - money2.twentyDollarCount,
    );
  }

  protected equalsCore(other: Money): boolean {
    return (
      this.oneCentCount === other.oneCentCount &&
      this.tenCentCount === other.tenCentCount &&
      this.quarterCount === other.quarterCount &&
      this.oneDollarCount === other.oneDollarCount &&
      this.fiveDollarCount === other.fiveDollarCount &&
      this.twentyDollarCount === other.twentyDollarCount
    );
  }
}
