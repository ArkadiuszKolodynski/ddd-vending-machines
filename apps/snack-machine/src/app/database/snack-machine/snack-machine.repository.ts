import { EntityManager, EntityRepository, LoadStrategy } from '@mikro-orm/core';
import { InjectRepository } from '@mikro-orm/nestjs';
import { Injectable, Provider } from '@nestjs/common';
import { SnackMachine } from '../../snack-machine/snack-machine';
import { SnackMachineRepository } from '../../snack-machine/snack-machine.repository.interface';
import { SnackMachineEntity } from './snack-machine.entity';
import { SnackMachineMapper } from './snack-machine.mapper';

@Injectable()
export class MikroOrmSnackMachineRepository implements SnackMachineRepository {
  constructor(
    @InjectRepository(SnackMachineEntity)
    private readonly snackMachineRepository: EntityRepository<SnackMachineEntity>,
    private readonly em: EntityManager,
  ) {}

  async findOne(): Promise<SnackMachine | null> {
    const snackMachineEntity = await this.snackMachineRepository.findOne(
      { id: { $exists: true } },
      {
        populate: ['slots', 'slots.snackPile', 'slots.snackPile.snack'],
        strategy: LoadStrategy.SELECT_IN, // FIXME: https://github.com/mikro-orm/mikro-orm/issues/4546
      },
    );

    if (!snackMachineEntity) {
      return null;
    }

    const snackMachine = SnackMachineMapper.toDomain(snackMachineEntity);
    return snackMachine;
  }

  async save(snackMachine: SnackMachine): Promise<void> {
    const snackMachineEntity = await this.snackMachineRepository.findOne(
      { id: snackMachine.id },
      {
        populate: ['slots'],
        strategy: LoadStrategy.SELECT_IN, // FIXME: https://github.com/mikro-orm/mikro-orm/issues/4546
      },
    );

    snackMachineEntity.assign(SnackMachineMapper.toPersistence(snackMachine));
    this.em.flush();
  }
}

export const SnackMachineRepositoryProvider: Provider = {
  provide: SnackMachineRepository,
  useClass: MikroOrmSnackMachineRepository,
};
