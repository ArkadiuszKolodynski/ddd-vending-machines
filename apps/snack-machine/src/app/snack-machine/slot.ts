import { Entity } from '@vending-machines/shared';
import { SnackMachine } from './snack-machine';
import { SnackPile } from './snack-pile';

export class Slot extends Entity {
  readonly snackMachine: SnackMachine;
  readonly position: number;
  snackPile: SnackPile;

  constructor(snackMachine: SnackMachine, position: number) {
    super();
    this.snackMachine = snackMachine;
    this.position = position;
    this.snackPile = SnackPile.Empty;
  }
}
