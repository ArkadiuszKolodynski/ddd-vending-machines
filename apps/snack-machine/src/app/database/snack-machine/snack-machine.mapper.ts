import { EntityData } from '@mikro-orm/core';
import { Money } from '@vending-machines/shared';
import Currency from 'currency.js';
import { Slot } from '../../snack-machine/slot';
import { SnackMachine } from '../../snack-machine/snack-machine';
import { SnackPile } from '../../snack-machine/snack-pile';
import { SnackMapper } from '../snack/snack.mapper';
import SnackMachineEntity from './snack-machine.entity';

export class SnackMachineMapper {
  private constructor() {}

  static toDomain(entity: SnackMachineEntity): SnackMachine {
    const snackMachine = new SnackMachine();
    return Object.assign(snackMachine, {
      id: entity.id,
      moneyInside: new Money(
        entity.money.oneCentCount,
        entity.money.tenCentCount,
        entity.money.quarterCount,
        entity.money.oneDollarCount,
        entity.money.fiveDollarCount,
        entity.money.twentyDollarCount,
      ),
      moneyInTransaction: new Currency(0),
      slots: entity.slots.toArray().reduce((slots, slotEntity) => {
        const snack = SnackMapper.toDomain(slotEntity.snackPile.snack);
        const slot = new Slot(snackMachine, slotEntity.position);
        Object.assign(slot, {
          id: slotEntity.id,
          snackPile: new SnackPile(snack, slotEntity.snackPile.quantity, new Currency(slotEntity.snackPile.price)),
        });
        slots.push(slot);
        return slots;
      }, []),
    });
  }

  static toPersistence(snackMachine: SnackMachine): EntityData<SnackMachineEntity> {
    return {
      money: { ...snackMachine.moneyInside },
      slots: snackMachine.slots
        .map((slot) => ({
          id: slot.id,
          position: slot.position,
          snackPile: {
            quantity: slot.snackPile.quantity,
          },
        }))
        .sort((a, b) => a.position - b.position),
    };
  }
}
