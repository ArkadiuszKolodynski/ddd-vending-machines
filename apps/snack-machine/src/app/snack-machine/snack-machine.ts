import { Injectable } from '@nestjs/common';
import { AggregateRoot, InvalidOperationException, Money } from '@vending-machines/shared';
import Currency from 'currency.js';
import { Slot } from './slot';
import { SnackPile } from './snack-pile';

@Injectable()
export class SnackMachine extends AggregateRoot {
  moneyInside: Money = Money.None;
  moneyInTransaction: Currency = new Currency(0);
  readonly slots: Slot[] = [new Slot(this, 1), new Slot(this, 2), new Slot(this, 3)];

  insertMoney(money: Money): void {
    const coinAndNotes = [Money.Cent, Money.TenCent, Money.Quarter, Money.Dollar, Money.FiveDollar, Money.TwentyDollar];
    if (!coinAndNotes.some((coinOrNote) => coinOrNote.equals(money))) {
      throw new InvalidOperationException('Invalid coin or note');
    }

    this.moneyInTransaction = this.moneyInTransaction.add(money.amount);
    this.moneyInside = Money.add(this.moneyInside, money);
  }

  returnMoney(): void {
    const moneyToReturn: Money = this.moneyInside.allocate(this.moneyInTransaction);
    this.moneyInside = Money.subtract(this.moneyInside, moneyToReturn);
    this.moneyInTransaction = new Currency(0);
  }

  canBuySnack(position: number): string {
    const snackPile = this.getSnackPile(position);

    if (snackPile.quantity === 0) {
      return 'The snack pile is empty';
    }

    if (this.moneyInTransaction.intValue < snackPile.price.intValue) {
      return 'Not enough money inserted to buy a snack';
    }

    if (!this.moneyInside.canAllocate(this.moneyInTransaction.subtract(snackPile.price))) {
      return 'Not enough change';
    }

    return String();
  }

  buySnack(position: number): void {
    const buySnackError = this.canBuySnack(position);
    if (buySnackError) {
      throw new InvalidOperationException(buySnackError);
    }

    const slot = this.getSlotByPosition(position);
    slot.snackPile = slot.snackPile.subtractOne();

    const change = this.moneyInside.allocate(this.moneyInTransaction.subtract(slot.snackPile.price));
    this.moneyInside = Money.subtract(this.moneyInside, change);
    this.moneyInTransaction = new Currency(0);
  }

  getSnackPile(position: number): SnackPile {
    return this.getSlotByPosition(position).snackPile;
  }

  loadSnacks(position: number, snackPile: SnackPile): void {
    this.getSlotByPosition(position).snackPile = snackPile;
  }

  loadMoney(money: Money): void {
    this.moneyInside = Money.add(this.moneyInside, money);
  }

  unloadMoney(): Money {
    if (this.moneyInTransaction.intValue > 0) {
      throw new InvalidOperationException('Cannot unload money during transaction');
    }
    const moneyToReturn = this.moneyInside;
    this.moneyInside = Money.None;
    return moneyToReturn;
  }

  private getSlotByPosition(position: number): Slot {
    const slot = this.slots.find((slot) => slot.position === position);
    if (!slot) {
      throw new InvalidOperationException(`Slot at position ${position} does not exist`);
    }
    return slot;
  }
}
