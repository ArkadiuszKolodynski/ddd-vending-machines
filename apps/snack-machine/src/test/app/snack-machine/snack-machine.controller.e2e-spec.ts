import { HttpStatus, INestApplication, Module } from '@nestjs/common';
import { CommandBus, QueryBus } from '@nestjs/cqrs';
import { Test } from '@nestjs/testing';
import request from 'supertest';
import { SnackMachineRepository } from '../../../app/snack-machine/snack-machine.repository.interface';

@Module({
  providers: [{ provide: SnackMachineRepository, useValue: { findOne: jest.fn(), save: jest.fn() } }],
  exports: [SnackMachineRepository],
})
class DatabaseModuleMock {}
jest.mock('../../../app/database/database.module', () => {
  return {
    DatabaseModule: DatabaseModuleMock,
  };
});

import { ValidationProvider } from '@vending-machines/shared';
import { SnackMachineController } from '../../../app/snack-machine/snack-machine.controller';

describe('SnackMachine - e2e', () => {
  const queryBusMock = { execute: () => ({ moneyInTransaction: '$1.00', moneyInside: '$1.00' }) };
  const testEndpoint = '/snack-machine';

  let app: INestApplication;

  beforeAll(async () => {
    const module = await Test.createTestingModule({
      controllers: [SnackMachineController],
      providers: [
        ValidationProvider,
        { provide: QueryBus, useValue: queryBusMock },
        { provide: CommandBus, useValue: { execute: jest.fn() } },
      ],
    }).compile();

    app = module.createNestApplication();
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('/PUT insert-money', () => {
    it('should return 200 OK', () => {
      return request(app.getHttpServer())
        .put(`${testEndpoint}/insert-money`)
        .send({ money: [1, 0, 0, 0, 0, 0] })
        .expect(HttpStatus.OK)
        .expect(queryBusMock.execute());
    });

    it('should return 400 BAD REQUEST when payload is not an array', async () => {
      const response = await request(app.getHttpServer())
        .put(`${testEndpoint}/insert-money`)
        .send({ money: 'not an array' })
        .expect(HttpStatus.BAD_REQUEST);

      expect(response.body.message).toMatchSnapshot();
    });

    it('should return 400 BAD REQUEST when payload is an array with <6 elements', async () => {
      const response = await request(app.getHttpServer())
        .put(`${testEndpoint}/insert-money`)
        .send({ money: [0, 0, 0, 0, 1] })
        .expect(HttpStatus.BAD_REQUEST);

      expect(response.body.message).toMatchSnapshot();
    });

    it('should return 400 BAD REQUEST when payload is an array with >6 elements', async () => {
      const response = await request(app.getHttpServer())
        .put(`${testEndpoint}/insert-money`)
        .send({ money: [0, 0, 0, 0, 0, 0, 1] })
        .expect(HttpStatus.BAD_REQUEST);

      expect(response.body.message).toMatchSnapshot();
    });

    it('should return 400 BAD REQUEST when payload is not an array of 0 and 1 integers', async () => {
      const response = await request(app.getHttpServer())
        .put(`${testEndpoint}/insert-money`)
        .send({ money: ['0', '0', '0', '0', '0', '1'] })
        .expect(HttpStatus.BAD_REQUEST);

      expect(response.body.message).toMatchSnapshot();
    });
  });

  describe('/POST buy-snack', () => {
    it('should return 200 OK', () => {
      return request(app.getHttpServer())
        .post(`${testEndpoint}/buy-snack`)
        .send({ position: 1 })
        .expect(HttpStatus.OK)
        .expect(queryBusMock.execute());
    });
  });

  describe('/POST return-money', () => {
    it('should return 200 OK', () => {
      return request(app.getHttpServer())
        .post(`${testEndpoint}/return-money`)
        .expect(HttpStatus.OK)
        .expect(queryBusMock.execute());
    });
  });

  describe('/GET money-in-machine', () => {
    it('should return 200 OK', () => {
      return request(app.getHttpServer())
        .get(`${testEndpoint}/money-in-machine`)
        .expect(HttpStatus.OK)
        .expect(queryBusMock.execute());
    });
  });
});
