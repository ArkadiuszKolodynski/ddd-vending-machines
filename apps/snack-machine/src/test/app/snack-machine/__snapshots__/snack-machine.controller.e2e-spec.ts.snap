// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnackMachineController - e2e GET /:id should return 400 BAD REQUEST when id is not a valid uuid 1`] = `"Validation failed (uuid v 4 is expected)"`;

exports[`SnackMachineController - e2e GET /:id should return 400 BAD REQUEST when id is not a valid uuid 2`] = `"Validation failed (uuid v 4 is expected)"`;

exports[`SnackMachineController - e2e GET /:id/money-in-machine should return 400 BAD REQUEST when id is not a valid uuid 1`] = `"Validation failed (uuid v 4 is expected)"`;

exports[`SnackMachineController - e2e POST / should return 400 BAD REQUEST when id is not a valid uuid 1`] = `
[
  "id must be a UUID",
]
`;

exports[`SnackMachineController - e2e POST /:id/buy-snack should return 400 BAD REQUEST when id is not a valid uuid 1`] = `"Validation failed (uuid v 4 is expected)"`;

exports[`SnackMachineController - e2e POST /:id/buy-snack should return 400 BAD REQUEST when payload is not an integer 1`] = `
[
  "position must not be greater than 3",
  "position must not be less than 1",
  "position must be an integer number",
]
`;

exports[`SnackMachineController - e2e POST /:id/buy-snack should return 400 BAD REQUEST when payload is not an integer between 1 and 3 1`] = `
[
  "position must not be greater than 3",
]
`;

exports[`SnackMachineController - e2e POST /:id/return-money should return 400 BAD REQUEST when id is not a valid uuid 1`] = `"Validation failed (uuid v 4 is expected)"`;

exports[`SnackMachineController - e2e PUT /:id/insert-money should return 400 BAD REQUEST when id is not a valid uuid 1`] = `"Validation failed (uuid v 4 is expected)"`;

exports[`SnackMachineController - e2e PUT /:id/insert-money should return 400 BAD REQUEST when payload is an array with <6 elements 1`] = `
[
  "money must contain at least 6 elements",
]
`;

exports[`SnackMachineController - e2e PUT /:id/insert-money should return 400 BAD REQUEST when payload is an array with >6 elements 1`] = `
[
  "money must contain no more than 6 elements",
]
`;

exports[`SnackMachineController - e2e PUT /:id/insert-money should return 400 BAD REQUEST when payload is not an array 1`] = `
[
  "money must contain no more than 6 elements",
  "money must contain at least 6 elements",
  "money must be an array",
  "each value in money must be one of the following values: 0, 1",
  "each value in money must be an integer number",
]
`;

exports[`SnackMachineController - e2e PUT /:id/insert-money should return 400 BAD REQUEST when payload is not an array of 0 and 1 integers 1`] = `
[
  "each value in money must be one of the following values: 0, 1",
  "each value in money must be an integer number",
]
`;

exports[`SnackMachineController - e2e PUT /:id/insert-money should return 400 BAD REQUEST when payload is not an array of integers 1`] = `
[
  "each value in money must be one of the following values: 0, 1",
  "each value in money must be an integer number",
]
`;
